Zend/zend.c

```diff
@@ -1377,6 +1380,68 @@ ZEND_API ZEND_COLD void zend_error_zstr_at(
 			break;
 		default:
 			/* Handle the error in user space */
+			op1_fake_ref = NULL;
+			op1_ref = NULL;
+			op2_ref = NULL;
+			opline = NULL;
+			if (EG(current_execute_data)) {
+				zend_execute_data *execute_data = EG(current_execute_data);
+
+				if (EX(func) && ZEND_USER_CODE(EX(func)->type)) {
+					opline = EX(opline);
+
+					if (opline) {
+						if (opline->op1_type & (IS_VAR|IS_TMP_VAR|IS_CV)
+						 && opline->opcode != ZEND_ROPE_ADD
+						 && opline->opcode != ZEND_ROPE_END) {
+							zval *zv = EX_VAR(opline->op1.var);
+							if (opline->op1_type == IS_VAR && Z_TYPE_P(zv) == IS_INDIRECT) {
+								/* Potentially, user error handler may destroy the enclosing array or object
```

Comment:
Only partially familiar with the indirect handling code - this seems to include opcodes that entail setting array fields by reference.

Is destroying the only concern? Is a erealloc that would cause `Z_INDIRECT_P(zv)` (found within arData/arPacked) to point to a different value possible, e.g. if the user error handler adds elements to the array/object and the size grows from 6 to 12 and the capacity needs to grow from 8 to 16 (or if it converts packed to hash)?

`zend_fetch_dimension_address` returns an IS_INDIRECT pointer to a zval within the arData/arPacked (for opcodes writing to a field of an array?)

```c
				if (EXPECTED(GC_DELREF(op1_fake_ref) == 1)) {
					/* This is a life fake reference. Convert it back to regular zval. */
					zend_execute_data *execute_data = EG(current_execute_data);

					zval *zv = EX_VAR(opline->op1.var);
					ZEND_ASSERT(Z_TYPE_P(zv) == IS_INDIRECT);
					zv = Z_INDIRECT_P(zv);
					ZEND_ASSERT(Z_TYPE_P(zv) == IS_REFERENCE && Z_REF_P(zv) == op1_fake_ref);
					ZVAL_UNREF(zv);
```

Reply:
IS_INDIRECT zval is used when we pass a pointer to zval embedded into the enclosing array or object. It's necessary only for "write" context. (When we write back to arrayor object element). The write operation on arrays and strings are possible only when they have refcount == 1. They might be "separated". 

The new code in ``zend_error`` converts ``op1`` into a fake reference if necessary and increments refcounter, so it cannot escape. But I afraid, you are right. The fact that we keep reference a live, dosn't necessary mean that the IS_INDIRECT zval holder is a live as well. It might be reallocated. It should be easy to rove this with a test case.